zaza/openstack/charm_tests/hacluster/tests.py

```diff
@@ -130,3 +134,67 @@ def test_930_scaleback(self):
         # zaza.model.block_until_unit_wl_status(other_principle_unit, 'active')
         zaza.model.block_until_all_units_idle()
         logging.debug('OK')
+
+
+class HaclusterScaleBackAndForthTest(HaclusterBaseTest):
+    """hacluster tests scaling back and forth."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running hacluster tests."""
+        super(HaclusterScaleBackAndForthTest, cls).setUpClass()
+        test_config = cls.test_config['tests_options']['hacluster']
+        cls._principle_app_name = test_config['principle-app-name']
+        cls._hacluster_charm_name = test_config['hacluster-charm-name']
+
+    def test_930_scaleback(self):
+        """Remove a unit, recalculate quorum and add a new one."""
+        principle_units = sorted(zaza.model.get_status().applications[
+            self._principle_app_name]['units'].keys())
+        self.assertEqual(len(principle_units), 3)
+        doomed_principle_unit = principle_units[0]
+        other_principle_unit = principle_units[1]
+        doomed_hacluster_unit = juju_utils.get_subordinate_units(
+            [doomed_principle_unit], charm_name=self._hacluster_charm_name)[0]
+        other_hacluster_unit = juju_utils.get_subordinate_units(
+            [other_principle_unit], charm_name=self._hacluster_charm_name)[0]
+
+        logging.info('Pausing unit {}'.format(doomed_hacluster_unit))
+        zaza.model.run_action(
+            doomed_hacluster_unit,
+            'pause',
+            raise_on_failure=True)
+        logging.info('OK')
+
+        logging.info('Removing {}'.format(doomed_principle_unit))
+        zaza.model.destroy_unit(
+            self._principle_app_name,
+            doomed_principle_unit,
+            wait_disappear=True)
+        logging.info('OK')
+
+        logging.info('Waiting for model to settle')
+        zaza.model.block_until_unit_wl_status(other_hacluster_unit, 'blocked')
+        zaza.model.block_until_unit_wl_status(other_principle_unit, 'blocked')
```

Comment:
This line is expecting that the principle unit will go into blocked state when there is no quorum, although that's not happening when the tests are run:

This is using a juju status from the crashdump captured by this run https://openstack-ci-reports.ubuntu.com/artifacts/test_charm_pipeline_func_full/openstack/charm-hacluster/741592/22/8108/index.html in this patch https://review.opendev.org/c/openstack/charm-hacluster/+/741592

```
$ cat juju_status.yaml | yq --tojson eval | jq '.applications.keystone.units."keystone/1"."workload-status".current' 
"active"
$ cat juju_status.yaml | yq --tojson eval | jq '.applications.keystone.units."keystone/2"."workload-status".current' 
"active"
$ cat juju_status.yaml | yq --tojson eval | jq '.applications.keystone.units."keystone/1".subordinates."hacluster/2"."workload-status".current'  
"blocked"
$ cat juju_status.yaml | yq --tojson eval | jq '.applications.keystone.units."keystone/2".subordinates."hacluster/0"."workload-status".current'  
"blocked"
```
In percona-cluster charm there is `min-cluster-size` that will make the principle unit to go into blocked state when there are not enough units, but for keystone there doesn't seem to something similar, so this line could be dropped or change 'blocked' to 'active'.

Reply:
This must be a change of behavior in the charm in the linked Gerrit review then, because the older test class `HaclusterScalebackTest` contains also this check and the current charm (before the review) is passing this check. Digging

Reply:
Ah no I understand: `HaclusterScalebackTest` isn't part of the gate of the hacluster charm yet, good catch!

Reply:
But then this contradicts https://github.com/openstack-charmers/zaza-openstack-tests/pull/369#issuecomment-775866792, where I noticed that the `keystone` units would become blocked and the whole point of this test is to prove that with the new feature, the `keystone` units will afterwards exit the blocked state.

If `keystone` doesn't go into blocked state in the first place, then this entire test doesn't prove anything. I'm now wondering if `keystone` may end up going sometimes in blocked state, sometimes not.